File: week9/src/911/sample/file-counter.c
/*
        file-counter.c -- ファイルに値を保存するカウンタ
        ~yas/syspro/file/file-counter.c
*/

#include <stdio.h>  /* stderr, fopen(), fread(), frwite(), fclose() */
#include <stdlib.h> /* exit(), strtol() */
#include <string.h> /* strcmp() */

int counter_load();
void counter_save(int counter);

#define DATA_FILENAME "file-counter-value.data"
char *comname = "";

void usage() {
    printf("Usage$ %s {get,inc,set newvalue} {args}\n", comname);
    printf("Data file: %s\n", DATA_FILENAME);
    exit(1);
}

int main(int argc, char *argv[]) {
    char *cmd;
    int counter, newvalue;
    comname = argv[0];
    if (argc <= 1) usage();
    cmd = argv[1];
    counter = 0;
    if (strcmp(cmd, "get") == 0) {
        counter = counter_load();
    } else if (strcmp(cmd, "inc") == 0) {
        counter = counter_load();
        counter++;
    } else if (strcmp(cmd, "set") == 0) {
        if (argc != 3) usage();
        newvalue = strtol(argv[2], 0, 10);
        counter = newvalue;
    } else {
        usage();
    }
    printf("%d\n", counter);
    if (strcmp(cmd, "get") != 0) counter_save(counter);
    return 0;
}

void counter_save(int counter) {
    FILE *f;
    f = fopen(DATA_FILENAME, "w");
    if (f == NULL) {
        perror(DATA_FILENAME);
        exit(2);
    }
    if (fwrite(&counter, sizeof(counter), 1, f) != 1) {
        perror("fwrite");
        fclose(f);
        exit(2);
    }
    fclose(f);
}

int counter_load() {
    FILE *f;
    int counter;
    f = fopen(DATA_FILENAME, "r");
    if (f == NULL) {
        return 0;
    }
    if (fread(&counter, sizeof(counter), 1, f) != 1) {
        perror("fread");
        fclose(f);
        exit(2);
    }
    fclose(f);
    return counter;
}

File: week10/src/sample/file-counter.py
#!/usr/bin/env python3
# 	ファイルに値を保存するカウンタ(Python)
# 	~yas/syspro/file/file-counter-python.rb
import sys

DATA_FILENAME = "file-counter-value.data"
comname = sys.argv[0]

def usage():
    print(f"Usage$ {comname} {{get,inc,set newvalue}}")
    print(f"Data file: {DATA_FILENAME}")
    exit(1)
    return

def main(argv):
    if len(argv) <= 1:
        usage()
    cmd = argv[1]
    counter = 0
    if cmd == "get":
        counter = counter_load()
    elif cmd == "inc":
        counter = counter_load()
        counter += 1
    elif cmd == "set":
        if len(argv) != 3:
            usage()
        newvalue = int(argv[2])
        counter = newvalue
    else:
        usage()
    print(counter)
    if cmd != "get":
        counter_save(counter)
    return

def counter_load():
    counter = 0
    try:
        with open(DATA_FILENAME, "rb") as f:
            counter_bytes = f.read(4)
            counter = int.from_bytes(counter_bytes, byteorder=sys.byteorder)
    except FileNotFoundError:
        pass
    return counter

def counter_save(counter):
    counter_bytes = counter.to_bytes(4, byteorder=sys.byteorder)
    with open(DATA_FILENAME, "wb") as f:
        f.write(counter_bytes)
    return

main(sys.argv)
